IPython/core/display.py

```diff
@@ -139,6 +141,117 @@ def _new_id():
+
+    def __init__(self):
+        self._objs = set()
+
+    def get_repr_mimebundle(self, obj, include=None, exclude=None, *, on_recursion=RecursiveObject):
+        """
+        return the representations of an object and associated metadata.
+
+        An given object can have many representation available, that can be defined
+        in many ways: `_repr_*_` methods, `_repr_mimebundle_`, or user-registered
+        formatter for types.
+
+        When given an object, :any:`get_repr_mimebundle` will search for the
+        various formatting option with their associated priority and return the
+        requested representation and associated metadata.
+
+
+        Parameters
+        ----------
+        obj : an objects
+            The Python objects to get the representation data.
+        include : list, tuple or set, optional
+            A list of format type strings (MIME types) to include in the
+            format data dict. If this is set *only* the format types included
+            in this list will be computed.
+        exclude : list, tuple or set, optional
+            A list of format type strings (MIME types) to exclude in the format
+            data dict. If this is set all format types will be computed,
+            except for those included in this argument.
+        on_recursion: callable
+            Return an object to compute the representation when recursion is
+            detected.
+
+
+        Returns
+        -------
+        (data, metadata) : named tuple of two dicts
+
+            - 0/.data: See :any:`DisplayFormatter.format`.
+            - 1/.metadata: See :any:`DisplayFormatter.format`
+
+        Note
+        ----
+
+        When :any:`get_repr_mimebundle` detect it is recursively called, it will
+        attempt to return the representation of :class:`RecursiveObject`. You
+        may register extra formatter for :class:`RecursiveObject`
+
+        If you are computing objects representation in a concurrent way (thread,
+        coroutines, ...), you should make sure to instanciate a
+        :class:`ReprGetter` object and use one per task to avoid race conditions.
+
+        See Also
+        --------
+            :func:`display`, :any:`DisplayFormatter.format`
+
+        """
+        from IPython.core.interactiveshell import InteractiveShell
+        fmt = InteractiveShell.instance().display_formatter.format
+        if id(obj) in self._objs:
+            return DataMetadata(fmt(on_recursion(obj), include=include, exclude=exclude))
+        else:
+            try:
+                self._objs.add(id(obj))
+                return DataMetadata(fmt(obj, include=include, exclude=exclude))
+            finally:
+                self._objs.discard(id(obj))
+
+# Expose this for convenience at the top level. Similar to what the random
+# module in python does. If you want to avoid weird behavior from concurrency:
+#   Instantiate your own.
+get_repr_mimebundle = ReprGetter().get_repr_mimebundle
```

Comment:
If `get_repr_mimebundle` instantiated an internal ReprGetter at call time rather than being a method on a global object, there shouldn't be a potential for race conditions.